download.py

```diff
@@ -17,55 +20,102 @@
+    )
 
 
 def http_download(url: str) -> DownloadResult:
     logger.info(f"Checking if {url} was already downloaded")
+    start_time = time.time()
+
+    provenance = {
+        "activity_name": "Input download",
+        "activity_description": "Downloads the input file from INPUT_URI",
+        "processing_time_ms": -1,
+        "start_time_unix": start_time,
+        "parameters": [],
+        "software_version": "",
+        "input_data": url,
+        "output_data": "",
+        "steps": [],
+    }
+
     fn = os.path.basename(urlparse(url).path)
-    input_file = os.path.join(input_file_dir, fn)
-    # TODO rethink how to return mime_type
-    _, extension = get_asset_info(input_file)
+    asset_id, extension = get_asset_info(fn)
+
+    input_file_dir = os.path.join(data_base_dir, asset_id)
+    input_file = os.path.join(data_base_dir, asset_id, fn)
     mime_type = extension_to_mime_type(extension)
 
-    # download if the file is not present (preventing unnecessary downloads)
-    start_time = time.time()
+    download_time = -1.0
+
     if not os.path.exists(input_file):
         logger.info(f"File {input_file} not downloaded yet")
-        # Create /data/input/ folder if not exists
+        # Create /data/<asset_id>/ folder if not exists
         if not os.path.exists(input_file_dir):
             logger.info(f"{input_file_dir} does not exist, creating it now")
             os.makedirs(input_file_dir)
         with open(input_file, "wb") as file:
             response = requests.get(url)
+            if response.status_code >= 400:
+                logger.error(
+                    f"Could not download url. Response code: {response.status_code}"
+                )
+                return DownloadResult(
+                    input_file,
+                    mime_type,
+                    dict(),
+                    download_time,
+                    f"Input failure: Could not download url. Response code: {response.status_code}",
+                )
             file.write(response.content)
             file.close()
-    download_time = (time.time() - start_time) * 1000  # time in ms
+        download_time = (time.time() - start_time) * 1000  # time in ms
+    else:
+        provenance["steps"] = ["Download skipped: input already exists"]
+
     return DownloadResult(
-        input_file, download_time, mime_type  # TODO add content_length
+        input_file, mime_type, provenance, download_time  # TODO add content_length
     )
 
 
 # e.g. s3://dane-asset-staging-gb/assets/2101608170158176431__NOS_JOURNAAL_-WON01513227.mp4
-def s3_download(s3_uri: str) -> Optional[DownloadResult]:
-    logger.info(f"Checking if {s3_uri} was already downloaded")
+def s3_download(url: str) -> DownloadResult:
+    logger.info(f"Checking if {url} was already downloaded")
+    start_time = time.time()
 
-    if not validate_s3_uri(s3_uri):
-        logger.error(f"Invalid S3 URI: {s3_uri}")
-        return None
+    provenance = {
```

Comment:
I agree with the dataclass idea, but not so much with the provenance steps being overkill. Since we use that for the Whisper worker, I think it would make sense to also use it for this worker, have something that is standard between all workers. Plus, I would only change this once we have a proper discussion within the team about what the format of the provenance should be. For now, we agreed to keep using this older format which seems to cover our needs, so no need to oversimplify here.